File: Templates/TypeEditor/src/Templates/Templates/VB.TypeEditor/CustomControlLibrary.Protocol/DataTransport/CustomPropertyStoreData.vb
Imports Microsoft.DotNet.DesignTools.Protocol.DataPipe

Namespace DataTransport

    ''' <summary>
    '''  Transport class to carry the content of the CustomPropertyStore 
    '''  from the DesignToolServer's server process to the client (Visual Studio)
    '''  and back.
    ''' </summary>
    <DebuggerDisplay("{GetDebuggerDisplay(),nq}")>
    Partial Public Class CustomPropertyStoreData
        Implements IDataPipeObject

        ' Backing fields for properties, since VB cannot have auto-properties with private setters.
        Private _someMustHaveId As String
        Private _dateCreated As DateTime
        Private _listOfStrings As String()
        Private _customEnumValue As Byte

        Public Sub New()
        End Sub

        Public Sub New(someMustHaveId As String, dateCreated As DateTime, listOfStrings As String(), customEnumValue As Byte)
            Me.SomeMustHaveId = someMustHaveId
            Me.DateCreated = dateCreated
            Me.ListOfStrings = listOfStrings
            Me.CustomEnumValue = customEnumValue
        End Sub

        Public Property SomeMustHaveId() As String
            Get
                Return _someMustHaveId
            End Get
            Private Set(ByVal value As String)
                _someMustHaveId = value
            End Set
        End Property

        Public Property DateCreated() As DateTime
            Get
                Return _dateCreated
            End Get
            Private Set(ByVal value As DateTime)
                _dateCreated = value
            End Set
        End Property

        Public Property ListOfStrings() As String()
            Get
                Return _listOfStrings
            End Get
            Private Set(ByVal value As String())
                _listOfStrings = value
            End Set
        End Property

        Public Property CustomEnumValue() As Byte
            Get
                Return _customEnumValue
            End Get
            Private Set(ByVal value As Byte)
                _customEnumValue = value
            End Set
        End Property

        Public Sub ReadProperties(ByVal reader As IDataPipeReader) Implements IDataPipeObject.ReadProperties
            SomeMustHaveId = reader.ReadString(NameOf(SomeMustHaveId))
            DateCreated = reader.ReadDateTimeOrDefault(NameOf(DateCreated))

            ListOfStrings = reader.ReadArrayOrNull(
                NameOf(ListOfStrings),
                Function(innerReader) innerReader.ReadString())

            CustomEnumValue = reader.ReadByte(NameOf(CustomEnumValue))
        End Sub

        Public Sub WriteProperties(ByVal writer As IDataPipeWriter) Implements IDataPipeObject.WriteProperties
            writer.Write(NameOf(SomeMustHaveId), SomeMustHaveId)
            writer.WriteIfNotDefault(NameOf(DateCreated), DateCreated)

            writer.WriteArrayIfNotNull(
                NameOf(ListOfStrings), ListOfStrings,
                Sub(innerWriter, value) innerWriter.Write(value))

            writer.Write(NameOf(CustomEnumValue), CustomEnumValue)
        End Sub

        Private Function GetDebuggerDisplay() As String
            Return $"ID: {SomeMustHaveId}; {NameOf(DateCreated)}: {DateCreated}"
        End Function
    End Class
End Namespace

File: Templates/TypeEditor/src/TemplateSolutions/CS.CustomTypeEditor/CustomControlLibrary.Protocol/DataTransport/CustomPropertyStoreData.cs
using System;
using System.Diagnostics;
using System.Diagnostics.CodeAnalysis;
using Microsoft.DotNet.DesignTools.Protocol.DataPipe;

namespace CustomControlLibrary.Protocol.DataTransport
{
    /// <summary>
    ///  Transport class to carry the content of the CustomPropertyStore 
    ///  from the DesignToolServer's server process to the client (Visual Studio)
    ///  and back.
    /// </summary>
    [DebuggerDisplay("{" + nameof(GetDebuggerDisplay) + "(),nq}")]
    public partial class CustomPropertyStoreData : IDataPipeObject
    {
        [AllowNull]
        public string SomeMustHaveId { get; private set; }

        public DateTime DateCreated { get; private set; }
        public string[]? ListOfStrings { get; private set; }
        public byte CustomEnumValue { get; private set; }

        public CustomPropertyStoreData()
        {
        }

        public CustomPropertyStoreData(
            string someMustHaveId,
            DateTime dateCreated,
            string[]? listOfStrings,
            byte customEnumValue)
        {
            // We use this extension method here, which works also for .NET Framework and
            // lower than .NET 7 versions; it's defined in GlobalUtilities.cs.
            SomeMustHaveId = someMustHaveId.OrThrowIfArgumentIsNullOrEmpty();
            DateCreated = dateCreated;
            ListOfStrings = listOfStrings;
            CustomEnumValue = customEnumValue;
        }

        public void ReadProperties(IDataPipeReader reader)
        {
            SomeMustHaveId = reader.ReadString(nameof(SomeMustHaveId));
            DateCreated = reader.ReadDateTimeOrDefault(nameof(DateCreated));

            ListOfStrings = reader.ReadArrayOrNull(
                nameof(ListOfStrings), 
                (reader) => reader.ReadString()!);

            CustomEnumValue = reader.ReadByte(nameof(CustomEnumValue));
        }

        public void WriteProperties(IDataPipeWriter writer)
        {
            writer.Write(nameof(SomeMustHaveId), SomeMustHaveId);
            writer.WriteIfNotDefault(nameof(DateCreated), DateCreated);

            writer.WriteArrayIfNotNull(
                nameof(ListOfStrings),
                ListOfStrings,
                (writer, value) => writer.Write(value));

            writer.Write(nameof(CustomEnumValue), CustomEnumValue);
        }

        private string GetDebuggerDisplay()
            => $"ID: {SomeMustHaveId}; {nameof(DateCreated)}: {DateCreated}";
    }
}
